spec/02-integration/04-admin_api/06-certificates_routes_spec.lua

```diff
@@ -156,31 +176,310 @@ describe("Admin API: #" .. kong_config.database, function()
           local json = cjson.decode(body)
           assert.is_string(json.cert)
           assert.is_string(json.key)
-          assert.same({ "foo.com", "bar.com" }, json.snis)
+          assert.same({ "foobar.com", "baz.com" }, json.snis)
         end
       end)
     end)
 
-    describe("GET", function()
-      it("retrieves all certificates", function()
+    describe("PUT", function()
+      local cert_foo, cert_bar
+      before_each(function()
+        dao:truncate_tables()
+
+        local res = assert(client:send {
+          method  = "POST",
+          path    = "/certificates",
+          body    = {
+            cert  = ssl_fixtures.cert,
+            key   = ssl_fixtures.key,
+            snis  = "foo.com",
+          },
+          headers = { ["Content-Type"] = "application/x-www-form-urlencoded" },
+        })
+        local body = assert.res_status(201, res)
+        cert_foo = cjson.decode(body)
+
+        local res = assert(client:send {
+          method  = "POST",
+          path    = "/certificates",
+          body    = {
+            cert  = ssl_fixtures.cert,
+            key   = ssl_fixtures.key,
+            snis  = "bar.com",
+          },
+          headers = { ["Content-Type"] = "application/x-www-form-urlencoded" },
+        })
+        local body = assert.res_status(201, res)
+        cert_bar = cjson.decode(body)
+      end)
+
+      it("creates a certificate if ID is not present in the body", function()
         local res = assert(client:send {
+          method  = "PUT",
+          path    = "/certificates",
+          body    = {
+            cert  = ssl_fixtures.cert,
+            key   = ssl_fixtures.key,
+            snis  = "baz.com",
+          },
+          headers = { ["Content-Type"] = "application/x-www-form-urlencoded" },
+        })
+
+        local body = assert.res_status(201, res)
+        local json = cjson.decode(body)
+
+        assert.is_string(json.cert)
+        assert.is_string(json.key)
+        assert.same({"baz.com"}, json.snis)
+
+        -- make sure we added an sni
+        res = assert(client:send {
+          method  = "GET",
+          path    = "/snis",
+        })
+
+        body = assert.res_status(200, res)
+        local json = cjson.decode(body)
+        assert.equal(3, #json.data)
+        assert.equal(3, json.total)
+
+        -- make sure we added our certificate
+        res = assert(client:send {
           method = "GET",
           path = "/certificates",
         })
 
         local body = assert.res_status(200, res)
         local json = cjson.decode(body)
-        assert.equal(1, json.total)
-        assert.equal(1, #json.data)
-        assert.is_string(json.data[1].cert)
-        assert.is_string(json.data[1].key)
-        assert.same({ "foo.com", "bar.com" }, json.data[1].snis)
+        assert.equal(3, #json.data)
+        assert.equal(3, json.total)
+      end)
+
+      it("returns 404 for a random non-existing id", function()
+
```

Comment:
Could we avoid starting functions/tests with a blank line? There are other such places in this patchset. Thanks!